src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -43,25 +41,56 @@
+  needs HOME and PATH for facter resolution, so leave those, along with GEM_HOME
+  which is necessary for third party extensions that depend on gems.
+
+  We need to set the JARS..REQUIRE variables in order to instruct JRuby's
+  'jar-dependencies' to not try to load any dependent jars.  This is being
+  done specifically to avoid JRuby trying to load its own version of Bouncy
+  Castle, which may not the same as the one that 'puppetlabs/ssl-utils'
+  uses. JARS_NO_REQUIRE was the legacy way to turn off jar loading but is
+  being phased out in favor of JARS_REQUIRE.  As of JRuby 1.7.20, only
+  JARS_NO_REQUIRE is honored.  Setting both of those here for forward
+  compatibility."
+  [env :- {schema/Str schema/Str}
+   gem-home :- schema/Str]
+  (let [whitelist ["HOME" "PATH"]
+        clean-env (select-keys env whitelist)]
+    (assoc clean-env
+      "GEM_HOME" (fs/absolute-path gem-home)
+      "JARS_NO_REQUIRE" "true"
+      "JARS_REQUIRE" "false")))
+
+(schema/defn ^:always-validate managed-load-path :- [schema/Str]
+  "Return a list of ruby LOAD_PATH directories built from the
+  user-configurable ruby-load-path setting of the jruby-puppet configuration
+  section and prepending specific JRuby directories necessary for third party
+  extensions extensions.  See SERVER-297 and SERVER-538 for more info."
+  [ruby-load-path :- [schema/Str]
+   jruby-home :- schema/Str]
+  (let [load-path (->> ruby-load-path
+                    (cons ruby-code-dir)
+                    (cons (str jruby-home "/lib/ruby/1.9"))
+                    (cons (str jruby-home "/lib/ruby/shared"))
+                    (cons (str jruby-home "/lib/ruby/1.9/site_ruby")))]
+    (->> load-path
+      (map fs/absolute-path)
+      (map fs/normalized-path)
+      (map str))))
+
 (defn prep-scripting-container
   [scripting-container ruby-load-path gem-home]
-  (doto scripting-container
-    (.setLoadPaths (cons ruby-code-dir
-                         (map fs/absolute-path ruby-load-path)))
-    (.setCompatVersion (CompatVersion/RUBY1_9))
-    (.setCompileMode RubyInstanceConfig$CompileMode/OFF)
-    ;; We need to set the JARS..REQUIRE variables in order to instruct
-    ;; JRuby's 'jar-dependencies' to not try to load any dependent jars.  This
-    ;; is being done specifically to avoid JRuby trying to load its own version
-    ;; of Bouncy Castle, which may not the same as the one that
-    ;; 'puppetlabs/ssl-utils' uses. JARS_NO_REQUIRE was the legacy way to turn
-    ;; off jar loading but is being phased out in favor of JARS_REQUIRE.  As of
-    ;; JRuby 1.7.20, only JARS_NO_REQUIRE is honored.  Setting both of those
-    ;; here for forward compatibility.
-    (.setEnvironment (merge {"GEM_HOME" gem-home
-                             "JARS_NO_REQUIRE" "true"
-                             "JARS_REQUIRE" "false"}
-                            jruby-puppet-env))))
+  (let [jruby-home (.getHomeDirectory scripting-container)]
+    (doto scripting-container
```

Comment:
Wow, that's a very unfortunate API design on their part :(

If they just had an interface that was implemented by both `ScriptingContainer` and `RubyInstanceConfig`... :(

The duck typing idea bugs me, quite a bit, but it does seem like our tests would catch it if the interface changed on either of the two objects in a future JRuby release.

I'm on the fence as to which I think is better, but I do agree with you that it would be a good idea to wrap the `compat-version` and `compile-mode` values in a function or a var and reference that from both of these initializers.